test/relayclient/RelayProvider.test.ts

```diff
@@ -0,0 +1,274 @@
+          }
+        ]
+      }
+    })
+
+    it('should call callback with error if relayTransaction throws', function () {
+      const badRelayClient = new BadRelayClient(true, false, dependencyTree, gsnConfig.relayHubAddress, gsnConfig.relayClientConfig)
+      const relayProvider = new RelayProvider(badRelayClient, underlyingProvider, gsnConfig)
+      const promisified = new Promise((resolve, reject) => relayProvider._ethSendTransaction(jsonRpcPayload, (error: Error | null): void => {
+        reject(error)
+      }))
+      return expect(promisified).to.be.eventually.rejectedWith(`Rejected relayTransaction call - should not happen. Reason: Error: ${BadRelayClient.message}`)
+    })
+
+    it('should call callback with error containing relaying results dump if relayTransaction does not return a transaction object', function () {
+      const badRelayClient = new BadRelayClient(false, true, dependencyTree, gsnConfig.relayHubAddress, gsnConfig.relayClientConfig)
+      const relayProvider = new RelayProvider(badRelayClient, underlyingProvider, gsnConfig)
+      const promisified = new Promise((resolve, reject) => relayProvider._ethSendTransaction(jsonRpcPayload, (error: Error | null): void => {
+        reject(error)
+      }))
+      return expect(promisified).to.be.eventually.rejectedWith('Failed to relay call. Results:')
+    })
+
+    it('should convert a returned transaction to a compatible rpc transaction hash response', async function () {
+      const gsnConfig = configureGSN({ relayHubAddress: relayHub.address })
+      const relayProvider = new RelayProvider(undefined, underlyingProvider, gsnConfig)
+      const response: JsonRpcResponse = await new Promise((resolve, reject) => relayProvider._ethSendTransaction(jsonRpcPayload, (error: Error | null, result: JsonRpcResponse | undefined): void => {
+        if (error != null) {
+          reject(error)
+        } else {
+          resolve(result)
+        }
+      }))
+      assert.equal(id, response.id)
+      assert.equal('2.0', response.jsonrpc)
+      // I don't want to hard-code tx hash, so for now just checking it is there
+      assert.equal(66, response.result.length)
+    })
+  })
+
+  // TODO: most of this code is copy-pasted from the RelayHub.test.ts. Maybe extract better utils?
+  describe('_getTranslatedGsnResponseResult', function () {
+    let relayProvider: RelayProvider
+    let testRecipient: TestRecipientInstance
+    let canRelayFailedTxReceipt: BaseTransactionReceipt
+    let innerTxFailedReceipt: BaseTransactionReceipt
+    let innerTxSucceedReceipt: BaseTransactionReceipt
+    let notRelayedTxReceipt: BaseTransactionReceipt
+    let misbehavingPaymaster: TestPaymasterConfigurableMisbehaviorInstance
+    let relayRequest: RelayRequest
+    // It is not stirctly necessary to make this test against actual tx receipt, but I prefer to do it anyway
+    before(async function () {
+      const TestRecipient = artifacts.require('TestRecipient')
```

Comment:
why one big fixture for all tests? (e.g. canRelayFailedTxReceipt belong to its test, not to the global `before`)

Reply:
They share 90% of setup code, and only differ in what I tell MisbehavingPaymaster to do. Would be tedious to separate into individual `before`s

Reply:
ok got it: its a new way of writing tests: the entire flow runs in the "before" section, and then multiple tests to check its side-effects.
(actually much better than re-running the same "fixture" for each test, and just do its tests)